### PR TITLE
Add class-hierarchy variant based on grapht

### PR DESCRIPTION
**Pull request by smowton**
This mirrors the existing class_hierarchyt, except that it inherits from
grapht and so is suitable for use with generic graph algorithms.

### CODE REVIEW

File: src/goto-programs/class_hierarchy.cpp [hunk 1]
```diff
@@ -47,6 +47,42 @@ void class_hierarchyt::operator()(const symbol_tablet &symbol_table)
...
+    if(it->second.is_type && it->second.type.id()==ID_struct)
+    {
+      const class_typet &class_type = to_class_type(it->second.type);
+
+      for(const auto &base : class_type.bases())
+      {
+        const irep_idt &parent = to_symbol_type(base.type()).get_identifier();
+        if(parent.empty())
+          continue;
+        add_edge(nodes_by_name.at(parent), nodes_by_name.at(it->first));
```
Comment (tautschnig) [inline src/goto-programs/class_hierarchy.cpp]: Nit pick: How about `if(!parent.empty()) add_edge ...`?

**Review by tautschnig — CHANGES_REQUESTED** [review]
A nit-pick and some whitespace changes, please.

File: src/goto-programs/class_hierarchy.h [hunk 2]
```diff
@@ -59,4 +60,36 @@ class class_hierarchyt
...
+ public:
+  /// Maps class identifiers onto node indices
+  typedef
+    std::unordered_map<irep_idt, node_indext, irep_id_hash> nodes_by_namet;
+
+ private:
+  /// Maps class identifiers onto node indices
+  nodes_by_namet nodes_by_name;
+
+ public:
```
Comment (tautschnig) [inline src/goto-programs/class_hierarchy.h]: My apologies, the key comments had gotten lost: weird 1-char indent and please merge the `public` sections, possibly by moving the `private` part to the bottom.

**Comment by smowton** [comment]
@tautschnig changes applied

File: src/goto-programs/class_hierarchy.h [hunk 3]
```diff
@@ -59,4 +60,35 @@ class class_hierarchyt
   void get_parents_trans_rec(const irep_idt &, idst &) const;
 };
 
+/// Class hierarchy graph node: simply contains a class identifier.
+class class_hierarchy_graph_nodet : public graph_nodet<empty_edget>
+{
+ public:
```
Comment (tautschnig) [inline src/goto-programs/class_hierarchy.h]: There's one 1-char indent left here.

**Comment by smowton** [comment]
fixd

**Comment by smowton** [comment]
@tautschnig fixed

**Review by peterschrammel — APPROVED** [review]
LGTM modulo clang-format

File: src/goto-programs/class_hierarchy.h [hunk 4]
```diff
@@ -59,4 +60,35 @@ class class_hierarchyt
   void get_parents_trans_rec(const irep_idt &, idst &) const;
 };
 
+/// Class hierarchy graph node: simply contains a class identifier.
+class class_hierarchy_graph_nodet : public graph_nodet<empty_edget>
+{
+  public:
```
Comment (tautschnig) [inline src/goto-programs/class_hierarchy.h]: Just noting that you've now got mixed indenting style for those.

**Review by tautschnig — APPROVED** [review]

**Comment by smowton** [comment]
I was evidently not having a good day with formatting. Fixed all clang-format complaints, will merge once Travis passes.

**Pull request merged**